pyshepseg/tilingstats.py

```diff
@@ -1129,6 +1311,178 @@ def calcPerSegmentSpatialStatsTiled(imgfile, imgbandnum, segfile,
+      numReadWorkers : int
+        Number of concurrent read workers that RIOS will use.
+      missingStatsValue : int
+        The value to fill in for segments that have no data.
+    
+    """
+    if not HAVE_RIOS:
+        raise PyShepSegStatsError('RIOS needs to be installed for this function')
+    
+    segds, segband, imgds, imgband = doImageAlignmentChecks(segfile, 
+        imgfile, imgbandnum)
+
+    attrTbl = segband.GetDefaultRAT()
+    existingColNames = [attrTbl.GetNameOfCol(i) 
+        for i in range(attrTbl.GetColumnCount())]
+        
+    # Note: may be None if no value set
+    imgNullVal = imgband.GetNoDataValue()
+    if imgNullVal is not None:
+        # cast to the same type we are using for imagery
+        # (GDAL records this value as double)
+        imgNullVal = tiling.numbaTypeForImageType(imgNullVal)
+    else:
+        # because we need to mask out parts of tiles not part of the
+        # segment we need the no data value set
+        raise PyShepSegStatsError("NoData value must be set on imgfile")
+        
+    if 'targetoptions' not in userFunc.__dict__:
+        raise PyShepSegStatsError("userFunc must be @jit or @njit decorated")
+        
+    if len(colNamesAndTypes) == 0:
+        raise PyShepSegStatsError("Must specify one or more columns")
+    
+    histColNdx = checkHistColumn(existingColNames)
+    segSize = attrTbl.ReadAsArray(histColNdx).astype(numpy.uint32)
+    
+    # close all files so they can be opened in RIOS
+    del attrTbl
+    del segband
+    del segds
+    del imgband
+    del imgds
+
+    controls = applier.ApplierControls()
+    controls.selectInputImageLayers([imgbandnum], 'imgfile')
+    # RIOS default is 256x256. This leads to too many incomplete
+    # segments and increases memory use dramatically. 
+    controls.setWindowSize(tiling.TILESIZE, tiling.TILESIZE)
+    
+    # now create a new temporary file for saving the new columns too
+    tempFileMgr = applier.TempfileManager(controls.tempdir)
+    tempKEA = tempFileMgr.mktempfile(prefix='pyshepseg_tilingstatsspatial_', suffix='.kea')
+    keaDriver = gdal.GetDriverByName('KEA')
+    tempKEADS = keaDriver.Create(tempKEA, 10, 10, 1, gdal.GDT_UInt32)
+    tempKEABand = tempKEADS.GetRasterBand(1)
+    tempKEAAttrTbl = tempKEABand.GetDefaultRAT()
+    # make same size as original
+    tempKEAAttrTbl.SetRowCount(segSize.size)
+    
+    # Create columns, as required (in temp file)
+    n_intCols, n_floatCols, statsSelection_fast = createUserColumnsSpatial(
+        colNamesAndTypes, tempKEAAttrTbl, [])
+        
+    inputs = applier.FilenameAssociations()
+    inputs.segfile = segfile
+    inputs.imgfile = imgfile
+    
+    # we don't actually write any outputs
+    outputs = applier.FilenameAssociations()
+
+    if numReadWorkers > 0:
+        print('using numReadWorkers', numReadWorkers)
+        conc = applier.ConcurrencyStyle(numReadWorkers=numReadWorkers,
+            readBufferPopTimeout=30, readBufferInsertTimeout=30)
```

Comment:
I think this was a mistake from when I was struggling to work out why it was timing out - turned out that was because I was asking for a non existent band number..
But this raises an interesting point - what happens if the user does need to adjust these timeouts for some reason? Maybe I should accept a whole `applier.ConcurrencyStyle` object instead?

Reply:
Hmmm... that is a very good point. So, in principle, just passing in a ConcurrencyStyle object allows the user full control, and they can work out what parameters actually give the best result in their case. It even allows playing with compute workers, which might be helpful in some cases, depending on the statistics being calculated. 

Yes, I think that is a good idea.

Reply:
Hmm varying the compute workers is not going to work - the writing back to the RAT all needs to happen on the same thread. So I'll probably have to add an assert that `numComputeWorkers == 0`...

Reply:
Oh yeah, true. I forgot that the write is happening inside the compute. Yes, require that to be zero, and computeWorkerKind to be CW_NONE.